components/layout/display_list/mod.rs

```diff
@@ -1660,13 +1669,26 @@ fn glyphs(
                     index: glyph.id(),
                     point,
                 };
-                glyphs.push(glyph);
+
+                // First glyph must never be elided. Otherwise, check if it's time to crop.
+                // The first character or atomic inline-level element on a line must be clipped rather than ellipsed.
+                // <https://www.w3.org/TR/css-ui-3/#text-overflow>
+                if !fragment.overflow_metadata.can_be_elided ||
+                    total_advance <= max_total_advance ||
```

Comment:
This logic only adds the sizes of text runs, and then compares with the containing block. Not considering borders and such is wrong:

```html
<!DOCTYPE html>
<div style="border: solid; width: 100px; white-space: pre; overflow: clip; text-overflow: ellipsis"><span style="border: 15px solid cyan">Lorem ipsum</span></div>
```

<img width="117" height="40" alt="Image" src="https://github.com/user-attachments/assets/b63aef17-a851-4512-8e56-bf6ca62a39c7" />